### PR TITLE
Add total Add/Del timeout

### PR DESCRIPTION
**Pull request by martinkennelly**
During scale testing, we found that if an instance
of whereabouts did not accomplish its task within a
period of time, the kubelet spawned an additional
whereabouts instance and the initial instance
remained active.

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

### CODE REVIEW

**Comment by martinkennelly** [comment]
This needs careful review before merging. I tested it and the fix works but there's a lot to consider here.
I am worried about context expiring while we have LE, and we are updating/updated backend.

**Comment by martinkennelly** [comment]
Part of me thinks we should not propagate the overall add/del context to when we have LE and we are acting on the backend. Let the request time out only handle this.

**Comment by martinkennelly** [comment]
@rthakur-est have you seen this with your testing?

**Comment by rthakur-est** [comment]
> @rthakur-est have you seen this with your testing?

I have seen context timing out and kubelet spawning a new CNI plugin instance. I might not have noticed that earlier instance was still active.

**Comment by martinkennelly** [comment]
> > @rthakur-est have you seen this with your testing?
> 
> I have seen context timing out and kubelet spawning a new CNI plugin instance. I might not have noticed that earlier instance was still active.

Turn on debug logs. Confirm to this PR when you see it. Thank you!

File: pkg/storage/kubernetes/ipam.go [hunk 1]
```diff
@@ -417,11 +418,11 @@ func IPManagementKubernetesUpdate(mode int, ipam *KubernetesIPAM, ipamConf where
 	var pool storage.IPPool
 	var err error
 
-	ctx, cancel := context.WithTimeout(context.Background(), storage.RequestTimeout)
-	defer cancel()
+	requestCtx, requestCancel := context.WithTimeout(ctx, storage.RequestTimeout)
```
Comment (maiqueb) [inline pkg/storage/kubernetes/ipam.go:421]: What about renaming these variables ? `ctxWithTimeout` or something like that.

File: pkg/types/types.go [hunk 2]
```diff
@@ -14,6 +15,7 @@ const (
 	DefaultLeaderLeaseDuration = 1500
 	DefaultLeaderRenewDeadline = 1000
 	DefaultLeaderRetryPeriod   = 500
+	AddDelTimelimit            = 2 * time.Minute
```
Comment (maiqueb) [inline pkg/types/types.go:18]: Could you help me understand why 2 minutes ? 

Writing the justification in the commit msg would probably be helpful for future us.

**Review by maiqueb — COMMENTED** [review]
I can't see nothing wrong with this approach. Awesome catch.

I got a couple of questions about the code.

According to the commit msg:

```
During scale testing, we found that if an instance
of whereabouts did not accomplish its task within a
period of time, the kubelet spawned an additional
whereabouts instance and the initial instance
remained active.
```

I'm assuming this behavior is expected, and clearly defined in the CNI spec. Linking to it might also be a good thing.

**Comment by rthakur-est** [comment]
> > > @rthakur-est have you seen this with your testing?
> > 
> > 
> > I have seen context timing out and kubelet spawning a new CNI plugin instance. I might not have noticed that earlier instance was still active.
> 
> Turn on debug logs. Confirm to this PR when you see it. Thank you!

From the logs, what I see is when context times out, the plugin is invoked by kubelet to delete any IP it might have assigned with the same container ID. And after that for the same pod, it invokes plugin to assign IP with different container ID. This seems to be expected behavior.
This approach works fine for allocation cases where kubelet retries multiple times. But on pod deletion, kubelet doesn't retry and CNI plugins are not expected to throw error during deletion. This results in stale IPs being left in ippools. I had tried clubbing the lease timeout and request timeout with 30sec total timer, but it left with lots of stale IPs on undeployment.
In your experiments, are you checking for duplicate IPs/stale IPs on deployment/undeployment?

File: pkg/storage/kubernetes/ipam.go [hunk 3]
```diff
@@ -378,20 +378,20 @@ func IPManagement(mode int, ipamConf whereaboutstypes.IPAMConfig, containerID st
 
 	go func() {
 		defer wg.Done()
-		ctx, cancel := context.WithCancel(context.Background())
 		res := make(chan error)
+		leCtx, leCancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (rthakur-est) [reply pkg/storage/kubernetes/ipam.go:382]: should it be the parent ctx instead of context.Background? 
How leCtx will honor the timeout set by top level ctx?

[the remark this answers is not in the document]
Comment (maiqueb) [reply pkg/storage/kubernetes/ipam.go:382]: Good catch @rthakur-est ! IMO, this should be `leCtx, leCancel := context.WithCancel(ctx)` (the parent ctx).

File: pkg/storage/kubernetes/ipam.go (hunk 1 above)
Comment (rthakur-est) [reply pkg/storage/kubernetes/ipam.go:421]: Since timeout check is still done on ctx instead of requestCtx, is it really required to have a new requestContext.

File: pkg/storage/kubernetes/ipam.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rthakur-est) [reply pkg/storage/kubernetes/ipam.go:382]: IMO, the top level context timeout is not required. Let the leader election run with indefinite timeout as is done currently and use releaseOnCancel from #145 carefully with request timeout ctx. If leader lease cancellation is done only after ippool update, that will ensure that we do not end up with duplicate IPs and faster relinquishing of lease will reduce the overall time.

File: pkg/storage/kubernetes/ipam.go [hunk 4]
```diff
@@ -359,13 +359,13 @@ func IPManagement(mode int, ipamConf whereaboutstypes.IPAMConfig, containerID st
 		defer wg.Done()
 		for {
 			select {
-			case <-stopM:
-				logging.Debugf("Stopped leader election")
-				result <- nil
+			case <-ctx.Done():
+				err = fmt.Errorf("time limit exceeded while waiting to become leader")
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:363]: The error is returned when the function ```IPManagement``` ends and it gets returned back to the CNI caller to log.
LMK if you want to log to something else as well. I think its fine.

File: pkg/storage/kubernetes/ipam.go [hunk 5]
```diff
@@ -359,13 +359,13 @@ func IPManagement(mode int, ipamConf whereaboutstypes.IPAMConfig, containerID st
...
+				stopM <- struct{}{}
 				return
 			case <-leader:
 				logging.Debugf("Elected as leader, do processing")
-				newip, err = IPManagementKubernetesUpdate(mode, ipam, ipamConf, containerID, podRef)
+				newip, err = IPManagementKubernetesUpdate(ctx, mode, ipam, ipamConf, containerID, podRef)
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:368]: We are - its getting returned by ```IPManagement```.

File: pkg/storage/kubernetes/ipam.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:382]: > You've renamed this variable to stop shadowing the context from the function's input parameter, right ?

Correct and to make it slightly easier to read the context is associated with.

> should it be the parent ctx instead of context.Background?

I thought long about this. I do not think it should be parent process. If the parent process killed the LE thread, then we cannot guarantee that we weren't currently acting on the backend in ```IPManagementKubernetesUpdate```. I want to exit out of that function first and then we can end the ```le.Run's``` context.

> How leCtx will honor the timeout set by top level ctx?

It is indirectly honoured. When top level ctx ends, first we return from:
```
	go func() {
		defer wg.Done()
		for {
			select {
			case <-ctx.Done():
				err = fmt.Errorf("time limit exceeded while waiting to become leader")
				stopM <- struct{}{}
				return
			case <-leader:
				logging.Debugf("Elected as leader, do processing")
				newip, err = IPManagementKubernetesUpdate(ctx, mode, ipam, ipamConf, containerID, podRef)
				stopM <- struct{}{}
				return
			case <-deposed:
				logging.Debugf("Deposed as leader, shutting down")
				result <- nil
				return
			}
		}
	}()
```

This allows us to be sure we have stopped working on the backend in ```IPManagementKubernetesUpdate```.
Returning from the go routine above causing the cancellation of ```le.Run()```s context. So, we honor the top level context indirectly.
I know its horribly confusing.. and that is how bugs are hidden.

> Good catch @rthakur-est ! IMO, this should be leCtx, leCancel := context.WithCancel(ctx) (the parent ctx).

I disagree. If we do this and if ```ReleaseOnCancel``` is enabled, we give give rise to a race condition in ```IPManagementKubernetesUpdate``` and some other whereabouts instance that can instantly get the lease lock. I want to make sure we are out of ```IPManagementKubernetesUpdate``` before stoppling LE (and releasing the lock).

> IMO, the top level context timeout is not required. Let the leader election run with indefinite timeout as is done currently and use releaseOnCancel from #145 carefully with request timeout ctx. If leader lease cancellation is done only after ippool update, that will ensure that we do not end up with duplicate IPs and faster relinquishing of lease will reduce the overall time.

The reason I have a top level context is to prevent multiple instances of wb when kubelet respawns another instance if the pods status field isn't populated. I also kept this outside the function so I could also have a timeout on cmdDel.

File: pkg/storage/kubernetes/ipam.go (hunk 1 above)
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:421]: > What about renaming these variables ? ctxWithTimeout or something like that.

I rather my naming tbh. Can I ask why you want to add the Timeout? Thank Miguel.

> Since timeout check is still done on ctx instead of requestCtx, is it really required to have a new requestContext.

The only good I can see that it does is make sure there isn't any dangling / hanging connections to the backend. @maiqueb  What do you think?

**Comment by martinkennelly** [comment]
> From the logs, what I see is when context times out, the plugin is invoked by kubelet to delete any IP it might have assigned with the same container ID. And after that for the same pod, it invokes plugin to assign IP with different container ID. This seems to be expected behavior.

Do you see the other instances of wb for that same pod still active?

> This approach works fine for allocation cases where kubelet retries multiple times. But on pod deletion, kubelet doesn't retry and CNI plugins are not expected to throw error during deletion.

It doesn't retry directly no, but I want to prevent multiple cmdDel wb instances living beyond a set time so that we are not hitting the backend and the lease object.

This results in stale IPs being left in ippools. I had tried clubbing the lease timeout and request timeout with 30sec total timer, but it left with lots of stale IPs on undeployment.
> In your experiments, are you checking for duplicate IPs/stale IPs on deployment/undeployment?

I was yes and it seemed fine after maybe 1000 allocations/deallocations. I didn't see any stale allocations after multiple IP allocations. I added a PR #148 to add it to this repo.

It seems we need more testing to see if this PR is ok.

File: pkg/storage/kubernetes/ipam.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (maiqueb) [reply pkg/storage/kubernetes/ipam.go:368]: Oh I see. My bad. 

Not fond of mutating variables defined on other threads routines, but I guess you just built upon the pre-existing code. It becomes hard to follow the code, imo.

**Comment by martinkennelly** [comment]
The coming weekend, I will create a scale test with this PR and using #148 to check for stale IPs or duplicate IPs. Will report back next week. @rthakur-est Besides stale IPs and duplicate IPs, is there anything else I should test for?
Id love to have a test to check each IP allocation in the IP pool with what is actually set on the pod..........

**Comment by rthakur-est** [comment]
> The coming weekend, I will create a scale test with this PR and using #148 to check for stale IPs or duplicate IPs. Will report back next week. @rthakur-est Besides stale IPs and duplicate IPs, is there anything else I should test for?
> Id love to have a test to check each IP allocation in the IP pool with what is actually set on the pod..........

The original issues were duplicate IP and stale IPs, so just check for these.

File: pkg/storage/kubernetes/ipam.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:382]: > Kubelet will respawn another instance only if it receives an error from cni plugin

Not only, if the CNI plugin doesn't respond within a set time, kubelet will check the "status" field, and try to create the pod again.

File: pkg/storage/kubernetes/ipam.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:368]: > Not fond of mutating variables defined on other threads routines, but I guess you just built upon the pre-existing code. It becomes hard to follow the code, imo.

For sure. Its not great to read. It should be refactored. We can do this (and sleep at night) when we have tests.

File: pkg/storage/kubernetes/ipam.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rthakur-est) [reply pkg/storage/kubernetes/ipam.go:382]: I wasn't aware of this. In that case, this timeout can be kept close to the timeout that kubelet has.

[the remark this answers is not in the document]
Comment (martinkennelly) [reply pkg/storage/kubernetes/ipam.go:382]: youll see it happen at scale. If you want to repro it locally, edit the wb code and put a long sleep in the cmdAdd path. Look at kubelet logs and when it recreates the pod, check the number of wb instances.

File: pkg/storage/kubernetes/ipam.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dougbtv) [reply pkg/storage/kubernetes/ipam.go:363]: I think it'll get printed when its returned, should be OK

**Review by dougbtv — APPROVED** [review]
@nicklesimba and I took a read through, and we think it looks good. Thanks for the careful considerations all around, and especially on the subcontext for the goroutine.
